File: scala/bob/src/main/scala/Bob.scala
object Bob {
  def response(statement: String): String ={
    val s = statement.trim
    val acronyms = Set("OK", "DMV")
    val isShouting = s.split("[^a-zA-Z]+").filter{ word =>
      word.toUpperCase == word && word.length > 1 && !acronyms.contains(word.toUpperCase)
    }.nonEmpty
    val isQuestion = s.endsWith("?")
    (isQuestion, isShouting) match {
      case (_, _) if s.isEmpty => "Fine. Be that way!"
      case (true, true) => "Calm down, I know what I'm doing!"
      case (true, false) => "Sure."
      case (false, true) => "Whoa, chill out!"
      case _ => "Whatever."
    }
  }
}

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

import Data.Char (isSpace, isLower, isUpper)
import Data.List (isPrefixOf)

responseFor :: String -> String
responseFor s
  | null . clean $ s = "Fine. Be that way!"
  | shouting s && asking s = "Calm down, I know what I'm doing!"
  | asking s = "Sure."
  | shouting s = "Whoa, chill out!"
  | otherwise = "Whatever."
  where
    clean = reverse . filter (not . isSpace)
    asking = (isPrefixOf "?") . clean
    shouting str = let exists = flip any $ str in (exists isUpper) && (not $ exists isLower)
